src/handlers/swap_quote_handlers.ts

```diff
@@ -0,0 +1,198 @@
+        });
+
+        let gas = ZERO;
+        if (takerAddress) {
+            gas = await this._estimiateGasOrThrowRevertErrorAsync({
+                to,
+                data,
+                from: takerAddress,
+                value,
+                gasPrice,
+            });
+        }
+
+        const apiSwapQuote: SwapQuoteResponse = {
+            // TODO not everything has 18 decimals
+            price: makerAssetAmount.dividedBy(takerAssetAmount).decimalPlaces(DEFAULT_TOKEN_DECIMALS),
+            to,
+            data,
+            value,
+            gasPrice,
+            protocolFee,
+            gas,
+            makerAssetAmount,
+            takerAssetAmount,
+            orders,
+        };
+        return apiSwapQuote;
+    }
+    public async getSwapQuoteAsync(req: express.Request, res: express.Response): Promise<void> {
+        // parse query params
+        const {
+            sellToken,
+            buyToken,
+            sellAmount,
+            buyAmount,
+            takerAddress,
+            slippagePercentage,
+            gasPrice,
+        } = parseGetSwapQuoteRequestParams(req);
+        const sellTokenAddress = findTokenAddress(sellToken, CHAIN_ID);
+        const buyTokenAddress = findTokenAddress(buyToken, CHAIN_ID);
+        const isEthSell = sellToken === ETH_SYMBOL;
+        try {
+            const swapQuote = await this.calculateSwapQuoteAsync(
+                buyTokenAddress,
+                sellTokenAddress,
+                buyAmount,
+                sellAmount,
+                takerAddress,
+                isEthSell,
+                slippagePercentage,
+                gasPrice,
+            );
+            res.status(HttpStatus.OK).send(swapQuote);
+        } catch (e) {
+            // If this is already a transformed error then just re-throw
+            if (isAPIError(e)) {
+                throw e;
+            }
+            switch (e.message) {
+                case SwapQuoterError.InsufficientAssetLiquidity:
+                    throw new ValidationError([
+                        {
+                            field: buyAmount ? 'buyAmount' : 'sellAmount',
+                            code: ValidationErrorCodes.ValueOutOfRange,
+                            reason: e.message,
+                        },
+                    ]);
+                case SwapQuoterError.AssetUnavailable:
+                    throw new ValidationError([
+                        {
+                            field: 'token',
+                            code: ValidationErrorCodes.ValueOutOfRange,
+                            reason: e.message,
+                        },
+                    ]);
+                default:
+                    throw new InternalServerError(e.message);
+            }
+        }
+    }
+
+    private async _estimiateGasOrThrowRevertErrorAsync(txData: Partial<TxData>): Promise<BigNumber> {
+        // Perform this concurrently
```

Comment:
Are these being done concurrently? Or is this supposed to be a future TODO?

Reply:
It is being performed concurrently, the first estimate gas is kicked off without an await (so non blocking). We then await on the `callAsync` result and throw that if it is a revert error.